Color preferences improvements

daschuer opened the pull request:
This improved the color preferences a bit. See:

![grafik](https://user-images.githubusercontent.com/1777442/77256750-6aabc380-6c70-11ea-9911-7219931b2bfa.png)

daschuer committed:
Improve strings in the color preferences pane

daschuer committed:
Allow to translate "Name" in the ColorPaletteEditor

daschuer committed:
Move button row below the palette and optimize column widths

daschuer committed:
Fill table with assigned cue numbers if no cue numbers are set

daschuer commented:
I would like to change this even more, to improve the workflow editing the palette.
![grafik](https://user-images.githubusercontent.com/1777442/77258624-14448200-6c7c-11ea-91c6-8a29dbb1ed3b.png)

Normally, the Palette Editor should be hidden. When pressing edit, the palette editor is shown. With the Palette selected for Tracks or Cues. The name is appended with "(Edited)"
There will be a "Save" and a "Remove" button. 
When pressing save, the palette is stored with "Name" and automatically selected for the for cues or tracks where Edit... was pressed   

Saving a palette with a name of a build in palette is not possible. We may pop up a dialog or create a unique automatically. 

What do you think?

Holzhaus commented on src/preferences/dialog/dlgprefcolorsdlg.ui:
```diff
@@ -62,7 +76,7 @@
         </property>
         <property name="text">
-         <string>Assign predefined colors to newly created hotcue points</string>
+         <string>Select consecutive palette colors for new hotcues</string>
```
The palette colors aren't necessarily consecutive as they can be in any order depending on the "Assign to Hotcue Index" column. However, I'd suggest we remove this option completely - see PR #2578,

daschuer replied:
I think it is very usable to have this option separate from the palette editor. 
Many users probably fine with the Mixxx colors and only like to select the default color or the color by number option. 
So I actually want the combobox back where we can select one or all colors from the current palette. 
I think you are used to the color by number feature, right? 
I like to have one default color. Currently the later is hidden and therefore unpredictable. 
We need to fix that as well.
Would it OK for you to bring that box back?

Holzhaus commented on src/preferences/colorpaletteeditormodel.cpp:
```diff
@@ -103,9 +103,15 @@ void ColorPaletteEditorModel::setColorPalette(const ColorPalette& palette) {
+        }
+    } else {
+        for (int i = 0; i < palette.size(); i++) {
+            hotcueColorIndicesMap.insert(i, i);
```
Now it's a bit cumbersome so set a single default color for a large palette because you'd have to remove all the other numbers in the index column by hand. It's also a bit confusing because we now have numbers for the built-in track color palettes, too. The previous situation was probably also confusing, so I'm not sure which one is better. If we keep this change I think we need a context menu item to unset all indices in the palette editor.

daschuer replied:
The idea was to have always the coloumn populated with the cue numbers and enable the use via the box. 
So you only need to edit the column if you are not fine with this. This should effect virtually no one, isn't it?

Holzhaus replied:
Let's say I have a palette with 10 colors. One of the colors is red. If I want all my new cues red, I have to set "assign to hotcue number" to 1 for red and an empty value for all others. Before, I could just double click next to the red color and type 1. Done.

Now I have to the the same and then delete the numbers from all other rows.

Holzhaus replied:
I think this should be move into the paletteditor(model) classes. If there are no assigned hotcue indices, we could display grey placeholder values in the "assign to hot cue numbers" table (the placeholder values are not saved) . As soon as you set one value eyplicitly, the placeholder values should disappear.

daschuer replied:
> Before, I could just double click next to the red color and type 1. Done.

Interesting. That was not self explaining to me. I have actually tried to add all cue numbers to red like 1, 2, 3 ... but it fails. 

A combobox with all palette colors + color by number for default cue color will be IMHO quite easy to understand compared to that.  

I like to avoid to bothering the user with the palette editor if he just want to select the default color.  

Does this work for you as well?

Holzhaus replied:
I agree that setting the default color via the palette editor causes too much friction, however let's keep this out of this PR as I have an idea for that that I'd like to try in a separate PR after this one has been merged.

Holzhaus replied:
> I think this should be move into the paletteditor(model) classes. If there are no assigned hotcue indices, we could display grey placeholder values in the "assign to hot cue numbers" table (the placeholder values are not saved) . As soon as you set one value explicitly, the placeholder values should disappear.

Can you implement this? The current change also prevent reordering the palette with no indices set. Also, palette could be big so you should check if `i` is less than the number of hotcues.

Holzhaus reviewed (changes requested):
Thanks for working on UI polish ;-)

This mostly looks good. The button placement is probably a matter of preference but I'm okay with that. I'm unsure about the hotcue index changes (see my inline comment).

Also, maybe we need another button in place of the Discard button. In your screenshot it looks like "Discard" is translated to "Schließen ohne zu Speichern" ("Close without Saving") for German users. This is unfortunate, but I guess we can't change it to "Verwerfen" ("Discard"), right? Because we don't actually close without saving - instead we delete the currently selected palette.

ronso0 commented:
> Normally, the Palette Editor should be hidden.

Why? It shows a preview of the palette.

daschuer commented:
>>    Normally, the Palette Editor should be hidden.

> Why? It shows a preview of the palette.

Because it's editable style is overwhelming. 
Editing the palette is a very rare edge use case, that should not be that prominent. 

Instead it would be better to show palette as it is shown in the color selector from the skin. 
That is finally what the most users want to do here. 

Instead of a combo box for selecting the palette, we can pup up a menu with all palettes shown as colors. That's way the user can compare the palettes and pick the best suitable.

ronso0 commented:
> Instead it would be better to show palette as it is shown in the color selector from the skin.
> That is finally what the most users want to do here.
> 
> Instead of a combo box for selecting the palette, we can pup up a menu with all palettes shown as colors. That's way the user can compare the palettes and pick the best suitable.

:+1:

Be-ing commented:
Maybe we should make the default color palettes immutable and create a new custom palette when the user saves any edit?

Holzhaus commented:
> Maybe we should make the default color palettes immutable and create a new custom palette when the user saves any edit?

The default color palette is immutable. If you change anything you need to enter a new name in the combo box before you can click save.

Be-ing commented:
> Instead it would be better to show palette as it is shown in the color selector from the skin.
That is finally what the most users want to do here.

I agree. I find the implementation of the palette editor as a table overcomplicated and I think it takes much more screen space than necessary. It would be more straightforward to show a grid of colors like the cue menu.

daschuer committed:
Introduce Compbobox for selecting the hotcue default color

daschuer committed:
Respect new hotcue default settings

daschuer committed:
Added a preview stripe for the selected palette.
Repopulate the default color selector when changing the palette.

daschuer committed:
Fix default default hotcue selection

daschuer commented:
Now it looks like this: 

![grafik](https://user-images.githubusercontent.com/1777442/77440040-1c213500-6de8-11ea-8716-8f84abb4f2dd.png)

daschuer commented:
I think it is a good step forward to merge it as it is. 
We may consider the palette editor usability and the remove of "[Controls]", "auto_hotcue_colors" in a follow up PR. 
However this is not critical for a release IMHO. The mass changing of orange to something else should however be included. 
Maybe we can focus on this?

Holzhaus commented:
> I think it is a good step forward to merge it as it is.

Yup, thank you. The palette previews are very nice. Can you also set a miniature version as icon for the combobox items? Not necessary for merge, but would look nice and users wouldn't need to select each palette individually to see the colors.

Also, can you please implement the [placeholder-style hotcue indices](https://github.com/mixxxdj/mixxx/pull/2589#discussion_r396140368) or just revert the auto-setting of hotcue indices before merge?

> The mass changing of orange to something else should however be included.
> Maybe we can focus on this?

We can, but as I said on #2547 I need help with the DB caching issues. If nobody figures out what causes them I cannot continue working on this. I already spent hours looking at the code but with my lack of the DB caching internals I can't spot the issue.

Holzhaus commented on src/preferences/dialog/dlgprefcolors.cpp:
```diff
@@ -47,12 +54,44 @@ void DlgPrefColors::loadSettings() {
+    comboBoxHotcueDefaultColor->setItemIcon(0, QIcon(pixmap));
+
+    for (int i = 0; i < hotcuePalette.size(); ++i) {
+        comboBoxHotcueDefaultColor->addItem(tr("Palette") +
```
"Palette N" is not the right term, why no use something like "Color N: #FF8000"?

Be-ing commented on src/preferences/dialog/dlgprefcolors.cpp:
```diff
@@ -94,7 +118,84 @@ void DlgPrefColors::slotApply() {
+
+QPixmap DlgPrefColors::drawPalettePreview(const QString& paletteName) {
+    foreach (const ColorPalette& palette, mixxx::PredefinedColorPalettes::kPalettes) {
+        if (paletteName == palette.getName()) {
```
This limits the drawing of previews to the predefined palettes. I think it would be better to move this function from DlgPrefColors to ColorPalette.

Holzhaus replied:
I would only do that if we need that preview somewhere else. Right now we don't. But we should use ColotPaletteSettings.getPalette() instead of iterating over the predefined palettes here.

Be-ing commented on src/preferences/dialog/dlgprefcolors.cpp:
```diff
@@ -94,7 +118,84 @@ void DlgPrefColors::slotApply() {
+    labelHotcuePalette->setPixmap(preview);
+
+    ColorPalette palette = mixxx::PredefinedColorPalettes::kDefaultTrackColorPalette;
+    foreach (const ColorPalette& pal, mixxx::PredefinedColorPalettes::kPalettes) {
```
Why use Qt `foreach` macro instead of C++ standard `for (x : y)`?

Holzhaus replied:
This loop can also be replaced with ColorPaletteSettings.getPalette I think

Be-ing commented on src/preferences/dialog/dlgprefcolors.cpp:
```diff
@@ -94,7 +118,84 @@ void DlgPrefColors::slotApply() {
+    }
+
+    bool autoHotcueColors =
+            m_pConfig->getValue(ConfigKey("[Controls]", "auto_hotcue_colors"), false);
```
How about we change the name of this ConfigKey to `hotcue_colors_by_number`? The current name is unclear. I don't think we should worry about breaking configurations considering we haven't released a beta yet and it's easy to change this.

daschuer replied:
In this case, I think we should get entirely rid of it.

Be-ing replied:
Then how do you propose to keep track of whether to color hotcues by number or not? I suppose we could use a special value like -1 for HotcueDefaultColorIndex, but I think it's more appropriate to keep a separate boolean ConfigKey.

daschuer replied:
This should be  done after we have considered @Holzhaus plans in a follow up PR.

Be-ing replied:
I don't understand what you mean. What would be done in that follow up PR? Are you talking about #2578? I think that PR is made obsolete by this one and we should close #2578 without merging.

Be-ing replied:
Ping. @Holzhaus what do you think?

Holzhaus replied:
The point of #2578 to remove auto_hotcue_colors as an option and always as behavior. Instead, I was planning to make the single default color combobox simply create a new palette on the fly where the hotcue color index is set to 1 for that color and unset it for all other colors. That would make it possible to always use palette.getColorForHotcue() conditionally without using if-else-ladders and checking CO objects. If this PR has been merged i'll look into rebasing #2578 and check if that is indeed a feasible solution.

Be-ing replied:
That's similar to what I [proposed](https://github.com/mixxxdj/mixxx/pull/2578#issuecomment-603000399) in #2578. However, I think implementing that would be more trouble than it is worth. What we have here is working and the implementation makes sense. I would just like to rename the `auto_hotcue_colors` ConfigKey to something more descriptive.

Be-ing replied:
I'd prefer to switch the name of a ConfigKey now than during beta.

Be-ing replied:
ping

Be-ing replied:
@daschuer what do you think?

daschuer replied:
The current name is not that bad IMHO
I like to keep it either as it is to not break existing configurations or dispose this parameter entirely and reflect the selected combobox value in one parameter.

Be-ing replied:
I do think `auto_hotcue_colors` is not a good name. On its own I have no idea what it means. `hotcue_colors_by_number` would make the code easier to understand. However, I'm not going to insist on blocking this PR over this.

Be-ing commented:
I think this is a good step forward. The new combobox to pick between color hotcues by number or picking a single color from the palette is more straightforward than the approach in #2578 to add a separate palette.

The palette editing workflow is still confusing and I like the proposal in https://github.com/mixxxdj/mixxx/pull/2589#issuecomment-602261422 . I also find it odd that I have to right click to add a color to the palette. I propose replacing that right click menu with "Add" and "Remove" buttons.

I also agree with @Holzhaus that it would be nice to put the palette previews as icons in the combox items. This would save some vertical screen space too.

Be-ing commented:
The orange color for the default is assigned to hotcue number 9. I thought we didn't want that assigned to a hotcue number?

Be-ing commented:
I think it's a bit confusing that we call the default palette "Mixxx Hotcue Colors" but it is used for both hotcues and the track colors. Maybe rename it to simply "Mixxx Colors"?

Holzhaus commented:
> I think it's a bit confusing that we call the default palette "Mixxx Hotcue Colors" but it is used for both hotcues and the track colors. Maybe rename it to simply "Mixxx Colors"?

I think @ronso0 wanted to design a track color palette which would be more gradient-like (like the Serato Track Color Palette for example), not just a few distinct colors. This can also happen in the beta phase.

daschuer committed:
Revert "Fill table with assigned cue numbers if no cue numbers are set"

This reverts commit 30aaea8a51e98dd242848c0a46f12f53c59c3c65.

daschuer committed:
Improve the labels of the colors in the combobox

daschuer committed:
introduce getHotcueColorPalette(const QString&) and make use of it.

daschuer committed:
Added Icons to the palette select boxes

daschuer commented:
![grafik](https://user-images.githubusercontent.com/1777442/77484511-06812f00-6e2b-11ea-9a2b-5589cd35726e.png)

daschuer commented:
> The orange color for the default is assigned to hotcue number 9. I thought we didn't want that assigned to a hotcue number?

When we select color by number, we don't have a explicit default color any more. 
However this does not matter because orange is the 9th color but we have only 8 hotcues.

daschuer commented:
Done. 

Seeing all the colors side by side, I think we have done a great job to make them distinguishable :-)

Be-ing commented:
> we have only 8 hotcues

No, some controllers have more than 8 hotcue buttons.

daschuer commented:
Is there a use case to reserve orange? Originally we have removed orange because it was eiter to close to yellow or read. 
This can be done with the cue number assignment column. 
Should we do it?

Be-ing commented:
I'm more concerned with the odd appearance with a palette of 9 colors on controllers with >8 hotcues than whether or not anyone wants to set a hotcue to orange.

daschuer commented:
Ah Ok. So We should use only 8 colors for auto assignment.

Be-ing commented:
> Ah Ok. So We should use only 8 colors for auto assignment.

Yes

daschuer committed:
Limit auto color asignment to 8

daschuer commented:
Done

Be-ing commented on src/util/color/colorpalette.cpp:
```diff
@@ -20,7 +20,8 @@ mixxx::RgbColor ColorPalette::colorForHotcueIndex(unsigned int hotcueIndex) cons
         // For hotcue n, get nth color from palette
-        colorIndex = hotcueIndex;
+        // But use only  8 to avoid odd apearances on Hercules P32 or similar
+        colorIndex = hotcueIndex % 8;
```
I did not mean that we should artificially limit palettes to 8 assigned hotcue colors. Users of controllers with >8 hotcues might want to set 16 distinct hotcue colors and I don't think we should prevent them from doing so. I meant that the default palette should only have 8 assigned hotcue colors.

Be-ing replied:
Fixed here: https://github.com/daschuer/mixxx/pull/50 
84a43ec

Be-ing commented on src/preferences/dialog/dlgprefcolors.cpp:
```diff
@@ -147,6 +162,23 @@ QPixmap DlgPrefColors::drawPalettePreview(const QString& paletteName) {
+    }
+    return QIcon();
+}
+
```
This isn't quite what I was expecting. Only showing 4 colors is somewhat misleading and arbitrary. I was thinking of moving the image generated by drawPalettePreview into the combobox icon. I think we have enough horizontal space to do it. Does QComboBox only support square icons? If so, then let's leave it how it is.

Be-ing replied:
I implemented it here: https://github.com/daschuer/mixxx/pull/50
![image](https://user-images.githubusercontent.com/9455094/77496425-f2cccd00-6e18-11ea-880b-322137875d26.png)
![Screenshot from 2020-03-24 21-54-42](https://user-images.githubusercontent.com/9455094/77497000-4390f580-6e1a-11ea-9862-72dc5c48cf8f.png)

Be-ing committed:
DlgPrefColors: use full palette preview for icons in comboboxes

Be-ing committed:
limit default hotcue palette to 8 colors

ronso0 commented:
my 2 cents about the UX:
as soon as the palette was edited, can we automatically enable the Discard & Save buttons and
**A** append "[edited]" to the palette name
or
**B** do it like in web forms: clicking Save does nothing but focusing the palette name lineedit with a highlighted border (a popup "Choose a new name to save the palette" would be too much?)
or
**C** keep the buttons disabled and just highlight the lineedit
?

ronso0 commented:
Shouldn't `X Close without saving` be `X Remove custom palette` ?

JaviVilarroig commented:
Hi. I find confusing that the palette configuration dialog allows me to change the Mixxx Hotcue Colors palette but then is not saving them. In the message it says "Color Palette "Mixxx Hotcue Colors" is a built-in palette, not writing to config!"
If they are not going to be saved, I must not be allowed to change them.
(Is this the right place to place my comments on the PR's? if not, sorry)

JaviVilarroig commented:
When I set a hotcue in a track i get the following error 4 times in the log:
```
Warning [Main]: QString::arg: Argument missing: WWidget { background-color: #ff8000; }, #ffc07f
```

daschuer commented:
> When I set a hotcue in a track i get the following error 4 times in the log:

This is an master issue already fixed in master. I will merge master to solve it here as well.

Holzhaus commented:
@JaviVilarroig You cannot overwrite the built-in palettes, but you can edit them and then type a different name into the QLineEdit. We should probably color the lineedit red when it contains the read-only palette name and the palette was edited.

daschuer committed:
Merge remote-tracking branch 'upstream/master' into color_preferences

ronso0 commented:
Yup, pretty much what I [suggested above](https://github.com/mixxxdj/mixxx/pull/2589#issuecomment-603942704)

daschuer committed:
Rename Intro Palette to Track Metadata und remove it from the user available list.

daschuer committed:
Inital hide the palette editor

daschuer committed:
Split Save as and Template Combobox

daschuer commented:
This is the result from a usability test with my son: 
![grafik](https://user-images.githubusercontent.com/1777442/77707284-f8671600-6fc4-11ea-951e-cd201bd3ac93.png)

![grafik](https://user-images.githubusercontent.com/1777442/77707249-de2d3800-6fc4-11ea-9b34-b8b75d0724c2.png)

ronso0 commented:
Thx. Quick feedback: (I'm currently not able to build)

the tiny square palette preview feels redundant
![image](https://user-images.githubusercontent.com/5934199/77707721-49c3d500-6fc6-11ea-8c81-1bf262f4a94f.png)

Did you include that drop-down preview from @Be-ing?
When closed, I also find that much more appealing than a preview _below_ each combobox.
![image](https://user-images.githubusercontent.com/5934199/77707829-9b6c5f80-6fc6-11ea-8dd8-849b62dce24e.png)

How about making this the default view, 
with no palette selected (empty combobox, ` - - - `)
editor table hidden
![image](https://user-images.githubusercontent.com/5934199/77707964-fdc56000-6fc6-11ea-8736-62abd210676e.png)

daschuer commented:
I will include it once the review comments are fixed. 
@Be-ing: will you find time for it?

daschuer commented:
> How about making this the default view,
with no palette selected (empty combobox, - - -)

It turns out that the edit button is a requirement, during my usability test.

ronso0 commented:
> > How about making this the default view,
> > with no palette selected (empty combobox, - - -)
> 
> It turns out that the edit button is a requirement, during my usability test.

What irritates me is that **Edit** is huge and **Close** is kind of hidden.
Can we show it like this:
![colorEditor_open](https://user-images.githubusercontent.com/5934199/77710656-edb17e80-6fce-11ea-8481-0ff5c548ae97.jpg)

![colorEditor_close](https://user-images.githubusercontent.com/5934199/77710661-f1dd9c00-6fce-11ea-974b-b844ea290cd2.jpg)

+ a warning when closing with unsaved changes? (if it isn't there already)

Be-ing commented:
>  @Be-ing: will you find time for it?

Yes, not at the moment. If not tonight, then tomorrow I'll work on addressing your review comments.

Be-ing commented:
If we split the editor off to a separate window, we could reuse it in the cue menu so users don't have to go into the preferences.

JaviVilarroig commented:
> @JaviVilarroig You cannot overwrite the built-in palettes, but you can edit them and then type a different name into the QLineEdit and hit save. We should probably color the lineedit red when it contains the read-only palette name and the palette was edited.

Ok. I was not aware of that. My point is that there was no indication about that in the interface, only in the log. I'm testing as a user only.
I think that if something is immutable I must not be allowed to edit it. If this is too complex to implement at least a clear message or indication in the interface warning me that I will not be able to save my changes unless I create a new palette.
I will test again taking that information into account.
On the other hand I have been unable to discover how to use a different palette, not the default one, for my hotcues.

Holzhaus commented:
> I think that if something is immutable I must not be allowed to edit it. If this is too complex to implement at least a clear message or indication in the interface warning me that I will not be able to save my changes unless I create a new palette.

The current solution is very convenient because you can use any existing palette as a starting point for a new, custom palette. I think that when you edit a non-writable palette and didn't change the name yet, the save button should be disabled and the lineedit should be highlighted (e.g. red-ish background or something like that). The former is already implemented in the original version, I don't know if that's broken on this branch.

daschuer commented:
Ehm ... You are discussing the old state.
Please test my latest changes. All this should be solved.

Holzhaus commented on src/preferences/colorpaletteeditor.h:
```diff
@@ -21,25 +21,29 @@ class ColorPaletteEditor : public QWidget {
-    QPushButton* m_pResetButton;
+    QPushButton* m_pCloseButton;
+    QPushButton* m_pRemoveButton;
+    parented_ptr<QPushButton>(m_pResetButton);
```
```
In Datei, eingebunden von /home/jan/Projects/mixxx/cmake_build/mixxx-lib_autogen/include/preferences/dialog/ui_dlgprefcolorsdlg.h:22,
                 von /home/jan/Projects/mixxx/src/preferences/dialog/dlgprefcolors.h:7,
                 von /home/jan/Projects/mixxx/src/preferences/dialog/dlgprefcolors.cpp:1:
/home/jan/Projects/mixxx/src/preferences/colorpaletteeditor.h:48:30: Warnung: unnötige Klammern in Deklaration von »m_pResetButton« [-Wparentheses]
   48 |     parented_ptr<QPushButton>(m_pResetButton);
      |   
```

JaviVilarroig commented:
> Ehm ... You are discussing the old state.
> Please test my latest changes. All this should be solved.

Now the palette editor is OK. The interface is very intuitive. Thanks :)

Holzhaus reviewed (changes requested):
Thanks, however I encountered some issues.

1. If I select another track or hotcue palette and then edit and save a palette in the paletteeditor (without hitting Apply first), the palette combobox is reset to the value it had when the dialog window was opened.
2. If I select a different color as default color, then switch to another hotcue palette without hitting apply first, the default color combobox is reset to the value it had when the dialog window was opened
3. Did you consider to show the whole palette inside the combobox, not below, as @Be-ing proposed? https://github.com/mixxxdj/mixxx/pull/2589#discussion_r397581979
3. IMHO the whole Palette Editor Workflow is much more complicated now:
  - The Save/Reset button is always enabled now, even if it's a no-op.
  - I cannot easily reset my palette by just hitting the reset button, I now need to select the matching palette first or risk losing my custom palette.
  - Using a predefined as a base for your own palette is now very complicated. I now need select a predefined palette in the comboxbox next to the reset button, then reset my existing palette to another palette, then make my palette changes and change the palette name before hitting save. If I forgot the last step I just lost my pre-selected custom palette. Before, I could just select a predefined palette, make changes, change the name and hit save. If I forgot the last step, the save button would be grayed out and I didn't risk losing a custom palette.
  - Resetting to different palettes is not really useful, because you can easily delete and create a new palette anyway. In my (obviously biased) opinion, coloring the palette name combobox background red if the palette was changed but name is not writable would have done the job much better instead of a second "Reset to Palette" combobox. Additionally, we could add a small warning icon next to a predefined name with tool tip "Palette is read-only".

JaviVilarroig commented:
The "Assign to Houtcue number" is not working for me. In a previous version it was OK but now has stopped working.
I understand that it allows me to automatically assign a color to to the hot cues by their number. And I have a memory to have tested it and find it working properly. But now it's simply ignored and all the hot cues get the default color.

Holzhaus commented:
> The "Assign to Houtcue number" is not working for me. In a previous version it was OK but now has stopped working.
> I understand that it allows me to automatically assign a color to to the hot cues by their number. And I have a memory to have tested it and find it working properly. But now it's simply ignored and all the hot cues get the default color.

Works for me. Please double-check.

Be-ing committed:
Merge remote-tracking branch 'upstream/pr/2589' into color_preferences

Be-ing committed:
DlgPrefColors: remove unused slotTrackPaletteChanged

Be-ing committed:
DlgPrefColors: fix black edge of palette preview pixmaps

daschuer committed:
Merge pull request #50 from Be-ing/color_preferences

DlgPrefColors: use full palette preview for icons in comboboxes

daschuer committed:
Use small edit button with elipsis

daschuer committed:
Remove superfluid parenthes

Be-ing commented:
> The "Assign to Houtcue number" is not working for me. In a previous version it was OK but now has stopped working.
I understand that it allows me to automatically assign a color to to the hot cues by their number. And I have a memory to have tested it and find it working properly. But now it's simply ignored and all the hot cues get the default color.

How about making the "Assign to Hotcue Number" column disabled (grayed out) when a single color is selected instead of assigning by hotcue number in the combobox above?

Be-ing commented:
I agree with @Holzhaus that the workflow for editing a palette is overcomplicated. I like the idea [proposed at the beginning of this PR](https://github.com/mixxxdj/mixxx/pull/2589#issuecomment-602261422) with the Edit buttons next to the palette selection comboboxes. I think that would be more straightforward.

daschuer committed:
Added message box when closing Palette Editor wit unsaved changes

daschuer committed:
Keep temoraray selections when updateding custom paletts

daschuer committed:
Keep temorary selected default color after chanigng palette

daschuer committed:
Improve gray out state of save and reset button

daschuer commented:
> 1.  ...

Done

> 2. ...

Done

> 3. ...

Done

> 4. IMHO the whole Palette Editor Workflow is much more complicated now:

> *  The Save/Reset button is always enabled now, even if it's a no-op.

Fixed

> *  I cannot easily reset my palette by just hitting the reset button, I now need to select the matching palette first or risk losing my custom palette.

Fixed 

 > * Using a predefined as a base for your own palette is now very complicated. I now need select a predefined palette in the comboxbox next to the reset button, then reset my existing palette to another palette, then make my palette changes and change the palette name before hitting save. If I forgot the last step I just lost my pre-selected custom palette. Before, I could just select a predefined palette, make changes, change the name and hit save. If I forgot the last step, the save button would be grayed out and I didn't risk losing a custom palette.

IMHO The situation is now improved. You can allays reset back to you previous save custom palette, or just not save your change. Using two custom palettes is possible, but I don't see a many use case for this. So I think we should focus on editing one custom palette, which I IMOH also improved now. 

> Resetting to different palettes is not really useful, because you can easily delete and create a new palette anyway. In my (obviously biased) opinion, coloring the palette name combobox background red if the palette was changed but name is not writable would have done the job much better instead of a second "Reset to Palette" combobox. Additionally, we could add a small warning icon next to a predefined name with tool tip "Palette is read-only".

This was the original behavior which mad me do this poll request. I was confused by it's behavior of a single combo box and thought the whole sing was broken. I guess this was also the experience of @JaviVilarroig 
My usability test also confirms that the users want to create just one custom palette in the first place. 
This works now in in the expected way.

daschuer commented:
> How about making the "Assign to Hotcue Number" column disabled (grayed out) when a single color is selected instead of assigning by hotcue number in the combobox above?

The palette editor is out of context of the current selections. Connection these unrelated thing willbe confusing. 

> I agree with @Holzhaus that the workflow for editing a palette is overcomplicated. I like the idea proposed at the beginning of this PR with the Edit buttons next to the palette selection comboboxes. I think that would be more straightforward.

I think you have a fundamental different design in mind. This would be also nice, but requires a mayor refactoring. 

Instead of dealing with different custom color palettes, we can keep just one "current hotcue palette" and one "current track palette". Selecting a new prepared palette, would be resetting the custom palette to the new palette. You can at any time edit the palette. 
This means  we have two edit buttons at the combo boxes above. 

Mixing these two concepts lead to confusion. So we should stick with one. 
Our current concept is more flexible, but slightly harder to understand.

Holzhaus commented:
> This was the original behavior which mad me do this poll request. I was confused by it's behavior of a single combo box and thought the whole sing was broken. I guess this was also the experience of @JaviVilarroig

I think previously it wasn't apparent that you need to change the name to save the palette if the palette name is read-only. Coloring the combobox red (and possibly additionally showing a warning icon next to it) would have sufficed.

> My usability test also confirms that the users want to create just one custom palette in the first place.
> This works now in in the expected way.

So making both a custom Hotcue color palette and a Track Color Palette is out of scope? I think this is a relevant use case. After making my first palette, it's extremely non-obvious how to create a second one. Why do we even have a combobox if the user isn't supposed to have more than one custom palette?

I think the previous behaviour was much more straightforward.

daschuer commented:
In the case we would have two edit buttons.
One for the Hotcue and one for the track color palette. Instead of n-custom palettes, we have a two fully customizable current palettes. That's would be exactly the way the key notation GUI works.

You may argue to easy loose custom edits, but we had never complains about it. Less features, more usability.

In the usability test with my son, it turns out that we need to separate the reset combo box from the name box in the curren proposal. The combination of two features to one was the main source of confusion.

Did you actually test the latest state?
All stumbling blocks from your list are resolved. So please test again.

Actually, think as developer, we are now biased to find the best usable solution.

@JaviVilarroig, @Ronso0
Would you mind to check the usability again? What do you think?

daschuer committed:
Improve naming and use int as index like Qt

ronso0 commented:
Okay, I played with a little. A few remarks:

I don't want to be warned about unsaved palette when I just opened the editor without having changed anything.

What about
* grey out the Save button to be greyed out until I entered a name
* Save button next to the palette name
* a static Open / Close button like

![prefColors_save2](https://user-images.githubusercontent.com/5934199/77828002-67558380-7119-11ea-9c5d-81c0b00962a1.jpg)

ronso0 commented:
Maybe **Remove Palette** can also go into the [_Name..._] [Save] row

Be-ing commented:
> I think previously it wasn't apparent that you need to change the name to save the palette if the palette name is read-only. Coloring the combobox red (and possibly additionally showing a warning icon next to it) would have sufficed.

This is an easy problem to solve if the Edit button is put next to the palette selection comboboxes. Just append " (Edited)" to the name of the palette when opening it in the editor.

Be-ing commented:
> Instead of dealing with different custom color palettes, we can keep just one "current hotcue palette" and one "current track palette". Selecting a new prepared palette, would be resetting the custom palette to the new palette. You can at any time edit the palette.

Putting "Edit" buttons next to the palette selection comboboxes does not require limiting the number of custom palettes.

This Open/Close button is IMO awkward. It doesn't make much sense to open the palette editor for a palette that isn't being used. The workflow requires creating a custom palette, then saving it, then selecting it in the palette selection comboboxes to use it. It would be more straightforward to put the Edit button to open the palette editor next to the palette selection comboboxes.

> The palette editor is out of context of the current selections. Connection these unrelated thing willbe confusing.

@JaviVilarroig was [just confused](https://github.com/mixxxdj/mixxx/pull/2589#issuecomment-604924716) by the the lack of connection.

ywwg commented:
I agree with the suggestion that the palette editor should be in a popup launched from this window from an "edit" button, since editing a palette is a specific action. So the user doesn't need to see it until they want to actually perform that action.  For instance, just looking at that screenshot, the "close" button is very confusing.  So in my idea the view in the preferences would be (with an extra "edit.." button):

![image](https://user-images.githubusercontent.com/888940/77828941-fdc37e00-70f4-11ea-8568-03851f0b7ceb.png)

And then the palette editor popup would be like:

![image](https://user-images.githubusercontent.com/888940/77828948-07e57c80-70f5-11ea-9066-710dfc775e0b.png)


not exactly but you get the idea.

ywwg commented:
in the main prefs window, maybe have a dropdown for custom palette:

Custom palettes:
[dropdown list selector] [edit...] [remove]

daschuer committed:
Made the Palette editor a context depending pop up box for the palette combo boxes

daschuer commented:
Now it looks like that, the editor holds the context from the combo boxes in the preferences dialog. 
You can select here a palette you want to edit. 
By default it is saved with an (edited) prefix. The original selection is kept for setting the palette to default. Resetting to a different default is not longer possible. 
Once you press save, the edited palette is selected in the combo box. 

Please test and review.   

![grafik](https://user-images.githubusercontent.com/1777442/77861722-fbf1db80-7216-11ea-8d3a-38847ca1b89c.png)
![grafik](https://user-images.githubusercontent.com/1777442/77861734-1fb52180-7217-11ea-8ffe-733fab908b69.png)

ywwg commented:
This looks good! Moving the editor to a separate dialog is much more clear.

The buttons at the bottom of the palette dialog should be standard should probably be "Reset, Discard, Close".  "Reset" reverts to defaults, "Discard" is close without saving, and "Close" saves and closes.  I'd prefer the "remove" button to be in the original pref pane next to "Edit...", and there should be a confirmation for that with "Cancel, Remove" as the buttons.

ronso0 commented:
This is much better! thx!

In the editor, can hotcue fields in edit mode be spinboxes? That would allow configuration with mouse only.

A little weird: I double-clicked a hotcue field, applied no change, closed it, and yet the Reset button got activated.

Be-ing commented:
This is getting better. Some remaining issues:
  * Putting add/remove colors in a right click menu is not easily discoverable. Let's move those actions to buttons.
  * The "Hotcue default color" combobox resets to nothing selected when the hotcue palette is changed. Mixxx behaves as if "By hotcue number" is selected. I think that behavior makes sense considering a hotcue index in one palette has no meaning if another palette is selected. However, the combobox should show "By hotcue number".

ronso0 commented:
As much as I see the benefit for having "Remove" next to edit, I fear this would increase the Pref window size too much, or crop palette names.
A Remove / Cancel dialog should be added nonetheless IMO

Be-ing commented on src/preferences/colorpaletteeditor.cpp:
```diff
@@ -114,46 +100,40 @@ ColorPaletteEditor::ColorPaletteEditor(QWidget* parent)
-            m_pSaveAsComboBox->addItem(paletteName);
-            m_pPaletteTemplateComboBox->addItem(paletteName);
+        if (paletteName == palette.getName()) {
+            saveName = paletteName + QChar(' ') + tr("(Edited)");
```
Should we leave the `(` and `)` out of the translated string?

daschuer commented:
> The buttons at the bottom of the palette dialog should be standard should probably be "Reset, Discard, Close". "Reset" reverts to defaults, "Discard" is close without saving, and "Close" saves and closes. 

These are the standard QT strings and I like them. "Close without saving" is much more descriptive than "Discard" IMHO. It is also unclear that "Close" also saves. 

> I'd prefer the "remove" button to be in the original pref pane next to "Edit...", and there should be a confirmation for that with "Cancel, Remove" as the buttons.

There is not enough space for that if you also consider French Translations. 

Remember that we talking here about an edge case from an edge case. This should be IMHO not that prominent in the GUI.

Be-ing commented on src/preferences/dialog/dlgprefcolors.cpp:
```diff
@@ -227,18 +217,62 @@ void DlgPrefColors::slotHotcuePaletteChanged(const QString& paletteName) {
+
+void DlgPrefColors::openColorPaletteEditor(
+        const QString& paletteName,
+        bool editHotcuePalette) {
```
I think the "Assign to hotcue number" column should be hidden in the dialog if this bool is false. It has no significance for the track colors.

daschuer replied:
This is an issue of the shared palettes model. A palette has always the column and it should therefore always visible and accessible, regardless from which combobox the editor was started.
This flag is only used to automatically select the newly edited palette. That's all.

Be-ing replied:
The shared model is no problem. We can just hide the column from the QTableView. I implemented it here: https://github.com/daschuer/mixxx/pull/52

daschuer replied:
I don't think we as should do this.

Be-ing replied:
What do others think? @Holzhaus, @ronso0?

Be-ing replied:
I don't have a strong opinion about this either way, but I think it would be a small improvement.

ronso0 replied:
Yeah let's hide it when a the editor is opened for track colors, even if it shows the same palettes as for hotcues.

Holzhaus replied:
Yes, I think we can hide the column. If a palette is used for track colors, the "assign to hotcue number" is of no concern. The only case where this could be confusing is when the same palette is used for both comboboxes (track color and hotcues). We have three options:

1. Always show the column (both for track colors and hotcue colors)
2. Hide the column for track color palettes
3. Hide the column for track color palettes unless the track color palette is also used for hotcues

I'm okay with any of these options.

daschuer replied:
I have just merged daschuer#52

Be-ing commented on src/preferences/dialog/dlgprefcolors.cpp:
```diff
@@ -227,18 +217,62 @@ void DlgPrefColors::slotHotcuePaletteChanged(const QString& paletteName) {
+
+    comboBoxHotcueColors->setCurrentText(hotcueColors);
+    comboBoxTrackColors->setCurrentText(trackColors);
+    comboBoxHotcueDefaultColor->setCurrentIndex(defaultColor);
```
```suggestion
    // Select "By hotcue number". A hotcue index in one palette has no meaning in another palette.
    comboBoxHotcueDefaultColor->setCurrentIndex(0);
```

daschuer replied:
It has, it has. If you wish to use the first index as default or if you have selected the "color by number" setting. 
This change was requested by @Holzhaus

daschuer committed:
Seti inital color for QColorDialog

daschuer commented:
> In the editor, can hotcue fields in edit mode be spinboxes? That would allow configuration with mouse only.

We may consider that. But that is quite some work for an edge case feature.

> A little weird: I double-clicked a hotcue field, applied no change, closed it, and yet the Reset button got activated.

Got it. The initial color of the color picker was not set.

daschuer committed:
Don't translate parentheses

daschuer commented:
Done.

daschuer committed:
Adjust default index when it is out set the new palette selected

daschuer commented:
Done

Be-ing commented on src/preferences/dialog/dlgprefcolors.h:
```diff
@@ -39,6 +39,10 @@ class DlgPrefColors : public DlgPreferencePage, public Ui::DlgPrefColorsDlg {
             bool editHotcuePalette);
     QPixmap drawPalettePreview(const QString& paletteName);
     QIcon drawPaletteIcon(const QString& paletteName);
+    void restoreComboBoxed(
```
```suggestion
    void restoreComboBoxes(
```
?

daschuer committed:
fix typo

daschuer commented:
Done

Be-ing committed:
ColorPaletteEditor: move Add/Remove Color to buttons

instead of a right click menu where they were not easily
discoverable

Be-ing commented:
Done: https://github.com/daschuer/mixxx/pull/51
![image](https://user-images.githubusercontent.com/9455094/77866175-86c2de00-71f7-11ea-96a0-a12567f9081a.png)

Be-ing committed:
ColorPaletteEditor: hide hotcue number column for track palette

It has no meaning in this context.

daschuer committed:
Merge pull request #51 from Be-ing/color_preferences

ColorPaletteEditor: move Add/Remove Color to buttons

JaviVilarroig commented:
Managed to test again. Now I feel very comfortable with the workflow. It feels very natural.
The hot cue by number now works prefect.
Thanks a lot!!

ronso0 commented:
Looking good!
Though I expected the Add Color button open a color picker.
Either that, or automatically select the new color in the table below.

Be-ing commented:
> I expected the Add Color button ... automatically select the new color in the table below.

ronso0 commented:
Anyone objects adding the Remove button next to Edit?
We can use the trash bin icon like in the hotcue menu.

Be-ing commented:
What do you mean? There is no Edit button.

ronso0 commented:
> What do you mean? There is no Edit button.
that one
![image](https://user-images.githubusercontent.com/5934199/77913656-efac6380-7294-11ea-9aec-5429be4b8b6e.png)

Be-ing commented:
Oh I was just looking at the popup dialog. Yeah it might make more sense to move the delete icon there.

daschuer commented:
I would prefere not to add any additional buttons there, in favour to focus the real use cases.

Deleting is a very edge case and on top unnecessary in most cases. 
It is only required during testing, what you are currently do.

Even edit is a rare used case.

I like to keep the GUI clean without any scroll bar in any language. Vertical scrolling is a greater hassle.

Apart from that I think a delete button is logically placed worng there, becsuse can be missleading. The uer may not expect that the palette is also used in the other combo box. 

So hiding it a bit is a good choice.

ronso0 commented:
Okay, agreed.

ywwg commented:
I'm ok with a delete icon in either place.

Be-ing commented:
I think this is ready for merge when we reach a conclusion about the `auto_hotcue_colors` ConfigKey [above](https://github.com/mixxxdj/mixxx/pull/2589#pullrequestreview-380552669).

It would help if others could share an opinion on https://github.com/mixxxdj/mixxx/pull/2589#pullrequestreview-383464975 but we could also discuss that in another PR after merging this.

ywwg commented:
I'll take a look but probably it's ok to merge if it's technically correct and then we can iterate on the design as needed

ywwg commented:
notes:

* for "add color / remove color" usually a + and - button at the lower right of the list box should be understandable.
* "remove palette" needs a confirmation dialog
* it is nonstandard to have a "save" button also close the box.  "OK" is more appropriate.
the order is a little weird.  "[reset][remove]              [close without saving][OK]" would be better

daschuer committed:
Resort the palette editor buttons. Replace "Save" with "OK"

daschuer committed:
Shrink add and remove button to + and - and move it below the table.

daschuer committed:
Improve selection behaviour during palette edit

daschuer committed:
Added a warning dialog when removing the palette.

daschuer commented:
![grafik](https://user-images.githubusercontent.com/1777442/77963439-a7656380-72dd-11ea-9098-101b3a0279c4.png)

Done.

Holzhaus commented:
Nice. I think you forgot to commit or push your button reordering though. This is how it looks for me on 4b3f9b49e182aa23e59a24666523e63bbbc226a7:
![2020-03-31-011831_846x668_scrot](https://user-images.githubusercontent.com/1834516/77970895-a805f600-72ed-11ea-9eeb-efbcac5058c8.png)

Holzhaus commented:
Another very minor nitpick: Can we use the `list-add` and `list-remove` icons instead of a regular "+" and "-"? If the icon is not available can can fall back to them.

daschuer commented:
I think we see now the behavior of different OSs wen using the QButtonBar.
Ubuntu for instsnce does not show the icons on the buttons.
I also wonder why you see discard, with a paper bin. This does not make sense at all in our context. Or does it?

Can you please Provide the suggested changes as a PR. Since I am not able to test it.

Be-ing commented:
Huh... it looks different from both of you for me:
![image](https://user-images.githubusercontent.com/9455094/77971780-35792400-72b5-11ea-92b6-ec4461e82a4a.png)

Be-ing commented on src/preferences/colorpaletteeditor.cpp:
```diff
@@ -170,17 +175,23 @@ void ColorPaletteEditor::slotTableViewDoubleClicked(const QModelIndex& index) {
+    m_pTableView->scrollToBottom();
+    //m_pTableView->selectionModel()->select(QItemSelection
+    //        m_pModel->index(m_pModel->rowCount() - 1, 0),
+    //        QItemSelectionModel::ClearAndSelect );
```
remove commented code

daschuer committed:
Remove commented code

ywwg commented:
This is looking really good! Can you reverse the + and -? That's what the usual UX convention is

Be-ing commented on src/preferences/colorpaletteeditor.h:
```diff
@@ -1,45 +1,53 @@
 #include "util/parented_ptr.h"
 
-// Widget for viewing, adding, editing and removing color palettes that can be
-// used for track/hotcue colors.
```
Restore this comment and add:
"Used by the Edit buttons in DlgPrefColors."

Be-ing commented on src/preferences/colorpaletteeditormodel.cpp:
```diff
@@ -102,9 +102,9 @@ void ColorPaletteEditorModel::setColorPalette(const ColorPalette& palette) {
+        int colorIndex = colorIndicesByHotcue.at(i);
         hotcueColorIndicesMap.insert(colorIndex, i);
     }
 
```
Add documentation to the header file:
"Model for the QTableView in ColorPaletteEditor"

Be-ing replied:
You forgot this one.

daschuer replied:
You have commented here the cpp file. In the header file we hae already comments: 
```
// Model that is used by the QTableView of the ColorPaletteEditor.
// Takes of displaying palette colors and provides a getter/setter for
// ColorPalette instances. 
```

Be-ing replied:
Sorry, I didn't see that in the diff for this PR so I missed it. But yeah there's nothing you need to do for this.

Be-ing commented on src/preferences/colorpalettesettings.h:
```diff
@@ -9,13 +9,17 @@ class ColorPaletteSettings {
             : m_pConfig(pConfig) {
```
Add documentation above:
"Saves ColorPalettes to and loads ColorPalettes from the mixxx.cfg file"

Be-ing commented on src/util/color/colorpalette.h:
```diff
@@ -6,10 +6,13 @@
 
```
Add documentation:
"An ordered list of colors that can be picked by WColorPicker, used for cue and track colors. Also used to map default colors to hotcues based on their hotcue number"

Be-ing reviewed (changes requested):
Add documentation for recently added classes

daschuer committed:
Swap "+" and "-" button

daschuer committed:
Added some comments

daschuer commented:
Done

Holzhaus commented on src/preferences/dialog/dlgprefcolors.cpp:
```diff
@@ -94,7 +143,156 @@ void DlgPrefColors::slotApply() {
+    if (paletteName == palette.getName()) {
+        for (int i = 0; i < palette.size() && i < 4; ++i) {
+            painter.setPen(mixxx::RgbColor::toQColor(palette.at(i)));
+            painter.setBrush(mixxx::RgbColor::toQColor(palette.at(i)));
```
Can you use `palette.colorForHotcueIndex()` instead? Right now, the icon may show colors that are not actually assigned to hotcues.

Holzhaus commented on src/preferences/dialog/dlgprefcolors.cpp:
```diff
@@ -94,7 +143,156 @@ void DlgPrefColors::slotApply() {
+    return QPixmap();
+}
+
+QIcon DlgPrefColors::drawPaletteIcon(const QString& paletteName) {
```
```suggestion
QIcon DlgPrefColors::drawHotcueColorByPaletteIcon(const QString& paletteName) {
```

Holzhaus reviewed (changes requested):
Looks good. I think this is pretty nice now.

I have some additional small changes in mind, but these can also go in a separate PR:
1. Use icons for +/- buttons if available
2. Add a Delete Icon to the "Remove Palette" button if available (currently it looks strange if the other buttons all have an icon, but this button does not)
3. Possibly replace the QDialogButtonBox with a HBoxLayout to get a consistent button order across all OSes (if desired)

daschuer committed:
Merge pull request #52 from Be-ing/color_preferences_hide_hotcue_number

ColorPaletteEditor: hide hotcue number column for track palette

daschuer committed:
Use the real hotcue colors when coloring the icon

daschuer commented:
Done. 

I have some objections to the proposed changes: 

> Use icons for +/- buttons if available

Sound good, in the first place however this may create a bunch of follow up issues: 
We just do make sure that the icon is actually drawn. Most of the Desktop environments have button icons disabled.  Currently the button resizes with the selected font, will this still work. 
Does any desktop environment have + and - icons. In case we do our  own icons, do the icons allways have enough contrast with the background used in the Desktop environment. Do they match to the overall desktop style? 

Given that only some user uses this dialog only a few times. I suggest to save you from all this work. 

> Add a Delete Icon to the "Remove Palette" button if available (currently it looks strange if the other buttons all have an icon, but this button does not)

Do you use KDE? I think this is one of the last environments using button Icons together with text.
It even features a setting to hide this button. So if you like to spend the time to add an edit icon on the edit button, please make sure it follows this setting.  

> Possibly replace the QDialogButtonBox with a HBoxLayout to get a consistent button order across all OSes (if desired)

I think this is a bad idea. All dialogs and even the common bar on the bottom of the preferences are using the QDialogButtonBox() this should feature a consistent user interface compared to other installed applications. It is part of the KDE style to have the Discard button right of OK. 
Chaining this would be a step back.

daschuer commented:
Merge?

ronso0 commented:
>     1. Use icons for +/- buttons if available
> 
>     2. Add a Delete Icon to the "Remove Palette" button if available (currently it looks strange if the other buttons all have an icon, but this button does not)

Thought about that as well, but setting icons is not as easy as in the skins or the Hotcue popup: we need to choose contarsting Icons and it's not predictable what background color OS themes use. Didn't look into that: can we use OS icons? At least Delete/Trash bin icon should be available

Holzhaus commented:
> Done.
> 
> I have some objections to the proposed changes:
> 
> > Use icons for +/- buttons if available
> 
> Sound good, in the first place however this may create a bunch of follow up issues:
> We just do make sure that the icon is actually drawn. Most of the Desktop environments have button icons disabled. Currently the button resizes with the selected font, will this still work.
> Does any desktop environment have + and - icons. In case we do our own icons, do the icons allways have enough contrast with the background used in the Desktop environment. Do they match to the overall desktop style?

Most DEs that don't have button icons for buttons *that also have a text label*. There are icons on buttons when there is no label, which is the case here.
I didn't look into it but I think `QIcon` offers a way to get the icon by name from the system's icon theme (`list-add`/`list-remove` according to the [FreeDesktop spec](https://specifications.freedesktop.org/icon-naming-spec/icon-naming-spec-latest.html)). If there are no icons, we should fall back to textual +/-.

> > Add a Delete Icon to the "Remove Palette" button if available (currently it looks strange if the other buttons all have an icon, but this button does not)
> 
> Do you use KDE? I think this is one of the last environments using button Icons together with text.
> It even features a setting to hide this button. So if you like to spend the time to add an edit icon on the edit button, please make sure it follows this setting.

No, I use [i3wm](https://i3wm.org/). I don't really care about the icon, but the inconsistency looks really bad (all other buttons *do* have an icon for me): ![Inconsistent button icons](https://user-images.githubusercontent.com/1834516/77970895-a805f600-72ed-11ea-9eeb-efbcac5058c8.png)

> 
> > Possibly replace the QDialogButtonBox with a HBoxLayout to get a consistent button order across all OSes (if desired)
> 
> I think this is a bad idea. All dialogs and even the common bar on the bottom of the preferences are using the QDialogButtonBox() this should feature a consistent user interface compared to other installed applications. It is part of the KDE style to have the Discard button right of OK.
> Chaining this would be a step back.

Yeah, I'm not sure about this either. I just thought that it's a bit confusing to have a different button order on different OSes (e.g. when we put screenshots into the manual). But I guess users will be able to live with that.

Be-ing commented:
I agree that using standard add/remove icons instead of "+" and "-" characters would look better.

Be-ing commented:
Icons changed: https://github.com/daschuer/mixxx/pull/55
![image](https://user-images.githubusercontent.com/9455094/78466623-e540fe00-76c8-11ea-8961-91aaf76abd07.png)

daschuer commented:
This is the way it looks at my machine now, somehow chunky :-/. Compared to @Be-ing s screenshot there seems to be a scaling issue. 
![image](https://user-images.githubusercontent.com/1777442/78470792-506fec00-772c-11ea-90fd-c745120034a8.png)

If I change to list-add-symbolic and list-remove-symbolic it looks like that: 
![image](https://user-images.githubusercontent.com/1777442/78470904-469ab880-772d-11ea-9124-5f378d029a10.png)
This is better, but there is not enough contrast. 

There is also he issue that this is a Linux only thing see: 
https://doc.qt.io/qt-5/qicon.html
> Note: By default, only X11 will support themed icons. In order to use themed icons on Mac and Windows, you will have to bundle a compliant theme in one of your themeSearchPaths() and set the appropriate themeName().

So seeing all, I think lets go with + - as text. This works at any OS without additional testing and also with screen readers. Giving that this dialog is not a main Mixxx use case, it is probably not worth to investigate this issue even further ,

Be-ing commented:
We ship Qt with its default built in theme on macOS and Windows. Does that not provide these icons?

Be-ing commented:
I'm going to go ahead and merge this now then open a new PR targeted at master for the +/- buttons.

Be-ing commented:
Good work everyone. We came up with a good solution together.

Be-ing commented:
@rryan can you add some documentation to the header files of the classes you modified and added? Just a sentence or two is often sufficient. If we don't insist on it with PRs, we'll never get the code documented.
